mlflow/tracking/context.py

```diff
@@ -0,0 +1,187 @@
+        Generate context-specific tags.
+
+        :return: dict of tags
+        """
+        pass
+
+
+class DefaultContext(ContextProvider):
+    def in_context(self):
+        return True
+
+    def tags(self):
+        return {
+            MLFLOW_SOURCE_NAME: _get_source_name(),
+            MLFLOW_SOURCE_TYPE: _get_source_type()
+        }
+
+
+class GitContext(ContextProvider):
+    def in_context(self):
+        return _get_source_version() is not None
+
+    def tags(self):
+        return {
+            MLFLOW_GIT_COMMIT: _get_source_version()
+        }
+
+
+class DatabricksNotebookContext(ContextProvider):
+    def in_context(self):
+        return databricks_utils.is_in_databricks_notebook()
+
+    def tags(self):
+        notebook_id = databricks_utils.get_notebook_id()
+        notebook_path = databricks_utils.get_notebook_path()
+        webapp_url = databricks_utils.get_webapp_url()
+        tags = {
+            MLFLOW_SOURCE_NAME: notebook_path,
+            MLFLOW_SOURCE_TYPE: SourceType.NOTEBOOK
+        }
+        if notebook_id is not None:
+            tags[MLFLOW_DATABRICKS_NOTEBOOK_ID] = notebook_id
+        if notebook_path is not None:
+            tags[MLFLOW_DATABRICKS_NOTEBOOK_PATH] = notebook_path
+        if webapp_url is not None:
+            tags[MLFLOW_DATABRICKS_WEBAPP_URL] = webapp_url
+        return tags
+
+
+class ContextProviderRegistry(object):
+    """Registry for context provider implementations
+
+    This class allows the registration of a context provider which can be used to infer meta
+    information about the context of an MLflow experiment run. Implementations declared though the
+    entrypoints `mlflow.context_provider` group can be automatically registered through the
+    `register_entrypoints` method.
+    """
+
+    def __init__(self):
+        self._registry = []
+
+    def register(self, context_provider_cls):
+        self._registry.append(context_provider_cls())
+
+    def register_entrypoints(self):
+        """Register tracking stores provided by other packages"""
+        for entrypoint in entrypoints.get_group_all("mlflow.context_provider"):
+            try:
+                self.register(entrypoint.load())
+            except (AttributeError, ImportError) as exc:
+                warnings.warn(
+                    'Failure attempting to register context provider "{}": {}'.format(
+                        entrypoint.name, str(exc)
+                    ),
+                    stacklevel=2
+                )
+
+    def __iter__(self):
+        return iter(self._registry)
+
+
+_context_provider_registry = ContextProviderRegistry()
+_context_provider_registry.register(DefaultContext)
+_context_provider_registry.register(GitContext)
+_context_provider_registry.register(DatabricksNotebookContext)
+
+_context_provider_registry.register_entrypoints()
+
+
+def _merge_tags(base, new):
+    return dict(itertools.chain(base.items(), new.items()))
+
+
+def resolve_tags(tags=None):
```

Comment:
I would document that this iterates through all registered context providers, and allows the provided `tags` to override any set there.